osquery/tables/system/windows/signature.cpp

```diff
@@ -0,0 +1,238 @@
+    break;
+  }
+
+  // There may be no signature, but we also end up here if the file is malformed
+  // or not accessible.
+  case TRUST_E_NOSIGNATURE: {
+    auto error_code = GetLastError();
+    if (error_code == TRUST_E_NOSIGNATURE ||
+        error_code == TRUST_E_SUBJECT_FORM_UNKNOWN ||
+        error_code == TRUST_E_PROVIDER_UNKNOWN) {
+      signature_info.result = SignatureInformation::Result::Missing;
+      exit_status = Status(0, "Ok");
+    } else {
+      exit_status =
+          Status(1, std::string("Failed to verify the file signature") + path);
+    }
+
+    break;
+  }
+
+  case ERROR_SUCCESS: {
+    signature_info.result = SignatureInformation::Result::Trusted;
+    exit_status = Status(0, "Ok");
+    break;
+  }
+
+  // Although this signature is perfectly valid, it wasn't explicitly trusted
+  // by the system administrator
+  case CRYPT_E_SECURITY_SETTINGS: {
+    signature_info.result = SignatureInformation::Result::Valid;
+    exit_status = Status(0, "Ok");
+    break;
+  }
+
+  case TRUST_E_SUBJECT_NOT_TRUSTED:
+  default: {
+    signature_info.result = SignatureInformation::Result::Untrusted;
+    exit_status = Status(0, "Ok");
+    break;
+  }
+  }
+
+  trust_provider_settings.dwStateAction = WTD_STATEACTION_CLOSE;
+  WinVerifyTrust(static_cast<HWND>(INVALID_HANDLE_VALUE),
+                 &authenticode_policy_provider,
+                 &trust_provider_settings);
+
+  return exit_status;
+}
+
+namespace tables {
+Status generateRow(Row& r, const std::string& path) {
+  r = {};
+
+  SignatureInformation signature_info;
+  auto status = querySignatureInformation(signature_info, path);
+  if (!status.ok()) {
+    std::stringstream error_message;
+    error_message << "Failed to verify the Authenticode signature for the "
+                     "following file: "
+                  << path << ". Error: " << status.getMessage();
+
+    return Status(1, error_message.str());
+  }
+
+  r << signature_info;
+  return Status(0, "Ok");
+}
+
+QueryData generateQueryResults(QueryContext& context) {
+  // The query must provide a predicate with constraints including path or
+  // directory. We search for the parsed predicate constraints with the equals
+  // operator.
+  auto paths = context.constraints["path"].getAll(EQUALS);
+  context.expandConstraints(
+      "path",
+      LIKE,
+      paths,
+      ([&](const std::string& pattern, std::set<std::string>& out) {
+        std::vector<std::string> patterns;
+        auto status =
+            resolveFilePattern(pattern, patterns, GLOB_ALL | GLOB_NO_CANON);
+        if (status.ok()) {
+          for (const auto& resolved : patterns) {
+            out.insert(resolved);
+          }
+        }
+        return status;
+      }));
+
+  QueryData results;
+  for (const auto& path_string : paths) {
+    Row r;
+    auto status = generateRow(r, path_string);
+    if (status.ok()) {
+      results.push_back(r);
+    } else {
+      LOG(ERROR) << status.getMessage();
```

Comment:
`ERROR` seems a bit intense for a failure to compute the signature of an executable, maybe this could be a `WARNING` instead? I think we try and reserve `ERROR` for something which would halt the daemon itself.